coderd/workspaceagents.go

```diff
@@ -925,6 +929,271 @@ func (api *API) postWorkspaceAppHealth(rw http.ResponseWriter, r *http.Request)
+		AccessToken:  gitAuthLink.OAuthAccessToken,
+		RefreshToken: gitAuthLink.OAuthRefreshToken,
+		Expiry:       gitAuthLink.OAuthExpiry,
+	}).Token()
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusOK, codersdk.WorkspaceAgentGitAuthResponse{
+			URL: redirectURL.String(),
+		})
+		return
+	}
+
+	if token.AccessToken != gitAuthLink.OAuthAccessToken {
+		// Update it
+		err = api.Database.UpdateGitAuthLink(ctx, database.UpdateGitAuthLinkParams{
+			ProviderID:        gitAuthConfig.ID,
+			UserID:            workspace.OwnerID,
+			UpdatedAt:         database.Now(),
+			OAuthAccessToken:  token.AccessToken,
+			OAuthRefreshToken: token.RefreshToken,
+			OAuthExpiry:       token.Expiry,
+		})
+		if err != nil {
+			httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+				Message: "Failed to update git auth link.",
+				Detail:  err.Error(),
+			})
+			return
+		}
+	}
+	httpapi.Write(ctx, rw, http.StatusOK, formatGitAuthAccessToken(gitAuthConfig.Type, token.AccessToken))
+}
+
+// Provider types have different username/password formats.
+func formatGitAuthAccessToken(typ codersdk.GitProvider, token string) codersdk.WorkspaceAgentGitAuthResponse {
+	var resp codersdk.WorkspaceAgentGitAuthResponse
+	switch typ {
+	case codersdk.GitProviderGitLab:
+		// https://stackoverflow.com/questions/25409700/using-gitlab-token-to-clone-without-authentication
+		resp = codersdk.WorkspaceAgentGitAuthResponse{
+			Username: "oauth2",
+			Password: token,
+		}
+	case codersdk.GitProviderBitBucket:
+		// https://support.atlassian.com/bitbucket-cloud/docs/use-oauth-on-bitbucket-cloud/#Cloning-a-repository-with-an-access-token
+		resp = codersdk.WorkspaceAgentGitAuthResponse{
+			Username: "x-token-auth",
+			Password: token,
+		}
+	default:
+		resp = codersdk.WorkspaceAgentGitAuthResponse{
+			Username: token,
+		}
+	}
+	return resp
+}
+
+func (api *API) gitAuthCallback(gitAuthConfig *gitauth.Config) http.HandlerFunc {
+	return func(rw http.ResponseWriter, r *http.Request) {
+		var (
+			ctx    = r.Context()
+			state  = httpmw.OAuth2(r)
+			apiKey = httpmw.APIKey(r)
+		)
+
+		_, err := api.Database.GetGitAuthLink(ctx, database.GetGitAuthLinkParams{
+			ProviderID: gitAuthConfig.ID,
+			UserID:     apiKey.UserID,
+		})
+		if err != nil {
+			if !errors.Is(err, sql.ErrNoRows) {
+				httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+					Message: "Failed to get git auth link.",
+					Detail:  err.Error(),
+				})
+				return
+			}
+
+			_, err = api.Database.InsertGitAuthLink(ctx, database.InsertGitAuthLinkParams{
+				ProviderID:        gitAuthConfig.ID,
+				UserID:            apiKey.UserID,
+				CreatedAt:         database.Now(),
+				UpdatedAt:         database.Now(),
+				OAuthAccessToken:  state.Token.AccessToken,
+				OAuthRefreshToken: state.Token.RefreshToken,
+				OAuthExpiry:       state.Token.Expiry,
+			})
+			if err != nil {
+				httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
```

Comment:
Seems a lot of these are 400s when they should be 500s but perhaps I'm missing something

Reply:
fixed em all up